backend/rest/rest.go

```diff
@@ -0,0 +1,212 @@
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/restic/restic/backend"
+)
+
+const connLimit = 10
+
+// Returns the url of the resource
+func restPath(url *url.URL, t backend.Type, name string) string {
+	location := url.String()
+	if !strings.HasSuffix(location, "/") {
+		location += "/"
+	}
+	if t == backend.Config {
+		return location + string(t)
+	}
+	return location + string(t) + "/" + name
+}
+
+type RestBlob struct {
+	b     *Rest
+	buf   *bytes.Buffer
+	final bool
+}
+
+func (rb *RestBlob) Write(p []byte) (int, error) {
+	if rb.final {
+		return 0, errors.New("blob already closed")
+	}
+	n, err := rb.buf.Write(p)
+	return n, err
+}
+
+func (rb *RestBlob) Read(p []byte) (int, error) {
+	return rb.buf.Read(p)
+}
+
+func (rb *RestBlob) Close() error {
+	rb.final = true
+	rb.buf.Reset()
+	return nil
+}
+
+// Finalize moves the data blob to the final location for type and name.
+func (rb *RestBlob) Finalize(t backend.Type, name string) error {
+	if rb.final {
+		return errors.New("Already finalized")
+	}
+
+	rb.final = true
+
+	// Check key does not already exist
+	resp, err := http.Get(restPath(rb.b.url, t, name))
+	if err == nil && resp.StatusCode == 200 {
+		return errors.New("Key already exists")
+	}
+
+	<-rb.b.connChan
+	_, err2 := http.Post(restPath(rb.b.url, t, name), "binary/octet-stream", rb.buf)
+	rb.b.connChan <- struct{}{}
+	rb.buf.Reset()
+	return err2
+}
+
+// Size returns the number of bytes written to the backend so far.
+func (rb *RestBlob) Size() uint {
+	return uint(rb.buf.Len())
+}
+
+// A simple REST backend.
+type Rest struct {
+	url      *url.URL
+	connChan chan struct{}
+}
+
+// Open opens the http backend at the specified url.
+func Open(url *url.URL) (*Rest, error) {
+	connChan := make(chan struct{}, connLimit)
+	for i := 0; i < connLimit; i++ {
+		connChan <- struct{}{}
+	}
+	return &Rest{url: url, connChan: connChan}, nil
+}
+
+// Location returns a string that specifies the location of the repository, like a URL.
+func (b *Rest) Location() string {
+	return b.url.Host
+}
+
+// Create creates a new Blob. The data is available only after Finalize()
+// has been called on the returned Blob.
+func (b *Rest) Create() (backend.Blob, error) {
+	blob := RestBlob{
+		b:   b,
+		buf: &bytes.Buffer{},
+	}
+	return &blob, nil
+}
+
+// Get returns an io.ReadCloser for the Blob with the given name of type t.
+func (b *Rest) Get(t backend.Type, name string) (io.ReadCloser, error) {
+	resp, err := http.Get(restPath(b.url, t, name))
+	if err == nil && resp.StatusCode != 200 {
+		err = errors.New("Not found")
+	}
+	return resp.Body, err
+}
+
+// GetReader returns an io.ReadCloser for the Blob with the given name of
+// type t at offset and length.
+func (b *Rest) GetReader(t backend.Type, name string, offset, length uint) (io.ReadCloser, error) {
+	rc, err := b.Get(t, name)
+	if err != nil {
+		return nil, err
+	}
+
+	n, errc := io.CopyN(ioutil.Discard, rc, int64(offset))
```

Comment:
Hm, I think this can be implemented with an HTTP range request. If I'm just interested in the last few bytes of a large blob, this code downloads and discards the rest. I'm aware that the S3 backend also uses this code, and I think we should check if S3 supports range request and fix the code there, too.